### PR TITLE
Refactor camerad

### PR DESCRIPTION
**Pull request by deanlee**
1.Add class CamraBuf  to handle buffers. remove a lot of redundant code.
2.Move all camera-related data to the corresponding files under /cameras.
3.Each camera uses its own callback function to process the frame, function name start with the prefix camera_process_.
4.The main.cc  only handles client connections.
5.Slightly modified visionbuf to meet the needs of refactor,and fixed the memory leak caused by visionbuf_to_cl

tested on QCOM

### CODE REVIEW

**Comment by ZwX1616** [comment]
Great PR! Is it possible that we remove all the unnecessary #ifdefs and move them to the individual files in cameras/ ? Ideally more generic main.cc the better

**Comment by deanlee** [comment]
That would be better.but it will cause a very large PR.
can we simply move the camera buffs into class in this PR?then create another PR to solve the rest.how do you think about this idea?

**Comment by deanlee** [comment]
ok i'll try

**Comment by deanlee** [comment]
It's basically done, @ZwX1616  how about this version?

**Comment by ZwX1616** [comment]
> It's basically done, only tested on QCOM. @ZwX1616 how about this version?

Looking good. I'll test it and let you know. Thanks!

**Comment by deanlee** [comment]
Continue to refactor based on @ZwX1616 's suggestions. thank you @ZwX1616 for your advice and encouragement on WeChat!

**Comment by deanlee** [comment]
Although there are still many things can be optimized, I think this PR has already done what I originally wanted to do.

File: selfdrive/camerad/cameras/camera_webcam.cc [hunk 1]
```diff
@@ -221,56 +213,72 @@ CameraInfo cameras_supported[CAMERA_ID_MAX] = {
...
 
-void cameras_run(MultiCameraState *s) {
-  set_thread_name("webcam_thread");
 
-  int err;
-  pthread_t rear_thread_handle;
-  err = pthread_create(&rear_thread_handle, NULL,
-                        rear_thread, &s->rear);
-  assert(err == 0);
+void camera_process_front(MultiCameraState *s, CameraState *c, int cnt, ) {
```
Comment (ihsakashi) [inline selfdrive/camerad/cameras/camera_webcam.cc:254]: tiny cleanup

**Comment by adeebshihadeh** [comment]
Continued in #2310

**Pull request closed without merging**